src/ServiceGetPositionFK.c

```diff
@@ -0,0 +1,148 @@
+    rosidl_runtime_c__String__fini(&g_messages_GetPositionFK.response.message);
+    geometry_msgs__msg__PoseStamped__fini(&g_messages_GetPositionFK.response.result);
+
+    MOTOROS2_MEM_TRACE_REPORT(svc_get_position_fk_fini);
+}
+
+//TODO(gavanderhoorn): refactor: create M+ Cart types <-> ROS types conversion lib
+static void MpCoord_To_GeomMsgsPose(MP_COORD const* const mp_coord, geometry_msgs__msg__Pose* const geom_pose)
+{
+    geom_pose->position.x = NANOMETERS_TO_METERS(mp_coord->x);
+    geom_pose->position.y = NANOMETERS_TO_METERS(mp_coord->y);
+    geom_pose->position.z = NANOMETERS_TO_METERS(mp_coord->z);
+    QuatConversion_MpCoordOrient_To_GeomMsgsQuaternion(
+        mp_coord->rx, mp_coord->ry, mp_coord->rz,
+        &geom_pose->orientation);
+}
+
+void Ros_ServiceGetPositionFK_Trigger(const void* request_msg, void* response_msg)
+{
+    GetPositionFK_Request* request = (GetPositionFK_Request*) request_msg;
+    GetPositionFK_Response* response = (GetPositionFK_Response*) response_msg;
+
+    Ros_Debug_BroadcastMsg("%s: entry", __func__);
+
+    //init response message
+    response->success = FALSE;
+    response->result_code = 0;
+    rosidl_runtime_c__String__assign(&response->message, "");
+    rosidl_runtime_c__String__assign(
+        &g_messages_GetPositionFK.response.result.header.frame_id, "");
+
+    Ros_Debug_BroadcastMsg("%s: requested: grp no: %u, tool no: %u",
+        __func__, request->group_no, request->tool_no);
+
+    //make sure a valid group is specified (group_no is unsigned, so cannot be < 0)
+    if (request->group_no >= g_Ros_Controller.numGroup)
+    {
+        //TODO(gavanderhoorn): shouldn't use "SelectionResultCodes" here
+        response->result_code = motoros2_interfaces__msg__SelectionResultCodes__INVALID_CONTROL_GROUP;
+        rosidl_runtime_c__String__assign(&response->message,
+            motoros2_interfaces__msg__SelectionResultCodes__INVALID_CONTROL_GROUP_STR);
+        goto DONE;
+    }
+
+    //make sure a valid tool is specified (tool_no is unsigned, so cannot be < 0)
+    if (request->tool_no >= MAX_VALID_TOOL_INDEX)
+    {
+        //TODO(gavanderhoorn): shouldn't use "SelectionResultCodes" here
+        response->result_code = motoros2_interfaces__msg__SelectionResultCodes__INVALID_SELECTION_INDEX;
+        rosidl_runtime_c__String__assign(&response->message,
+            motoros2_interfaces__msg__SelectionResultCodes__INVALID_SELECTION_INDEX_STR);
+        goto DONE;
+    }
+
+    //TODO(gavanderhoorn): convert request->joint_angles to moto angles
+    // Ros_CtrlGroup_ConvertRosUnitsToMotoUnits(..) always converts to pulses
+    long motoAnglePos[MP_GRP_AXES_NUM] = { 0 };
+
+    //call M+ API
+    MP_COORD fk_result = { 0 };
+    BITSTRING fig_ctrl = 0;
+    //TODO(gavanderhoorn): verify this is always in robot base frame
+    int res = mpConvAxesToCartPos(
+        request->group_no,
+        motoAnglePos,
+        request->tool_no,
+        &fig_ctrl, // unused
```

Comment:
The `fig-ctrl` can get tricky. Whether or not it's used is actually controlled by parameter `S2C430`.

As of YAS4.13 (Sept 2020), the default value of this parameter has changed. I'll need to do some more research before I say with any confidence what the values can/should be.

Reply:
I assumed `fig_ctrl` is an `[out]` argument, isn't it?

Reply:
You're right. I was thinking `Cart -> Angles`